More like angular

bencefarkas200 opened the pull request:
Végleges frontend, azokkal a változtatásokkal, amiket megbeszéltünk.

bencefarkas200 committed:
Service kialakítása (még nem működik)

bencefarkas200 committed:
no message

bencefarkas200 committed:
Mar minden mukodik ugy, mint regen, csak "angularosabb"!

bencefarkas200 committed:
(szerintem) végleges frontend

gy-szabo commented on front-end/db.json:
```diff
@@ -0,0 +1,16 @@
+{
```
úgy látom ez a fájl nincs használva sehol, ezért a repo-ban sincs rá szükség, törölhető

gy-szabo commented on front-end/src/app/app.component.css:
```diff
@@ -7,13 +7,27 @@ mat-form-field {
 }
 
-.task-done-container {
+.task-container {
```
sokban hasonlít a `.done-container`-re, ezért a közös stílusokat össze lehetne vonni, ami specifikus, az pedig maradna az adott osztályon.

így értem:
```css
.class1, .class2 {
  /* közös stílusok */
}

.class1 {
  /* spec stílusok */
}

.class2 {
  /* spec stílusok */
}
```

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -1,35 +1,126 @@
 import { Task } from 'src/model/Task';
+import { FormControl, FormGroup, ReactiveFormsModule } from '@angular/forms';
+import { CommonModule } from '@angular/common';
+import { TaskCardComponent } from './task-card/task-card.component';
```
olyan importok is maradtak bent, amik nincsenek használva. A legegyszerűbb megoldás az lenne, hogy az IDE-ben beállítod, hogy mentéskor törölje a nem használt importokat, így automatikusan törlődni fognak ezek

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -1,35 +1,126 @@
+import { TaskCardComponent } from './task-card/task-card.component';
+import { TaskService } from './task.service';
 
 @Component({
```
legyen standalone a komponens

gy-szabo commented on front-end/src/app/task-card/task-card.component.ts:
```diff
@@ -0,0 +1,42 @@
+import { TaskService } from '../task.service';
+import { Task } from 'src/model/Task';
+
+@Component({
```
legyen standalone a komponens

gy-szabo commented on javitando.txt:
```diff
@@ -0,0 +1,6 @@
+1. "app.component.html"-ben az inputokat formGroup-ba tenni
```
ezt a fájlt nem kell verzió követni. Jó ha neked van jegyzeted lokálisan, de a repóban erre nincs szükség

gy-szabo commented on front-end/src/app/task.service.ts:
```diff
@@ -0,0 +1,26 @@
+  submitTask(id: number, title: string, comment: string) {
+    if (title != '') {
+      this.db.push({ id: id, title: title, comment: comment });
+    } else alert("Title can't be empty!");
```
a materialnak van tök látványos megoldása ([snackbar](https://material.angular.io/components/snack-bar/overview)) ehelyett, az alert helyett kérlek használd ezt

gy-szabo replied:
másik ideillő dolog, hogy ha az else ág fut le, akkor is reset-eled a formot. Ez user szemszögből baj, mert pl ha a user 400 karaktert beírt title nélkül a commenthez, submitol egyet véletlen, akkor kap egy hibaüzenetet, és azt kezdheti majd előről bepötyögni a 400 karaktert

gy-szabo replied:
jött módosítás, de még mindig nem az igazi: jelenleg nem fut le az ellenőrzés, enged csak kommentet beírva taskokat felvinni. Ezt majd még átbeszéljük bővebben

gy-szabo commented on front-end/src/app/task.service.ts:
```diff
@@ -0,0 +1,26 @@
+import { jsDocComment } from '@angular/compiler';
```
nem használt import

gy-szabo commented on front-end/src/app/app.module.ts:
```diff
@@ -6,16 +6,19 @@ import { MatCardModule } from '@angular/material/card';
+    ReactiveFormsModule,
   ],
   providers: [],
   bootstrap: [AppComponent],
```
standalone komponenseket szeretnénk használni, ezért a modulos felosztásra nincs szükség, az app modulok nélkül működjön. Hivatalos doksiban van konkrét [module-standalone migrációról szóló guide](https://angular.io/guide/standalone-migration), hogy ez hogy lehet elérni

gy-szabo commented on front-end/src/app/app.module.ts:
```diff
@@ -6,16 +6,19 @@ import { MatCardModule } from '@angular/material/card';
 import { MatButtonModule } from '@angular/material/button';
 import { MatInputModule } from '@angular/material/input';
 import { FormsModule } from '@angular/forms';
+import { FormControl, FormGroup, ReactiveFormsModule } from '@angular/forms';
```
ezek közt is vannak 2 nem használt import

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -1,35 +1,126 @@
   doneArray: Task[] = [];
-  renderClearbtn = false;
+  doneArrayEmpty = true;
+  gotTaskArray = false;
```
a clean code betartása (ergo a kód olvashatósága) végett, a boolean típusú változókat tipikusan úgy nevezünk el, hogy már a névből kitalálható legyen, hogy 2 állapota van. 

Pl.: a `doneArrayEmpty` majdnem elfogadható, de a nevezéktani konvenciók helyett intuitívebb lenne az `isDoneArrayEmpty`. 

A `gotTaskArray` is majdnem jó, de a jelenlegi kódodban én inkább úgy nevezném el, hogy `shouldShowTaskArray`. Mivel egy bind-olva van egy HTML element-re, ezért a shouldShow tisztábban mutat arra, hogy valami HTML-es mókáért felel ez a változó

itt egy jó [cikk](https://medium.com/swlh/naming-conventions-101-for-developers-8997bb96fd60) ami taglalja ezeket

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -1,35 +1,126 @@
 
-  taskDelete(index: number) {
-    this.taskArray.splice(index, 1);
+  constructor() {
```
megint a clean code-dal jövök, mert abból sose elég :D Most arra akarok kitérni, hogy az osztályon belül legyen ez a séma, amit követünk:
1. konstansok
2. változók
3. konstruktorok
4. publikus függvények
5. privát függvények

Tehát legfelül legyenek a konstansok, változók, utánuk a konstruktor, és utánuk a függvények, ne pedig össze-vissza. Így minden osztály egységesen fog kinézni, és könnyebb lesz eligazodni rajtuk.

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -1,35 +1,126 @@
   styleUrls: ['./app.component.css'],
 })
 export class AppComponent {
+  taskService = inject(TaskService);
```
az `inject()` helyett a `taskService` a konstruktorban kapjon értéket. Valahogy így:
```js
  constructor(private readonly taskService: TaskService) {
    // ...
  }
```

gy-szabo commented on front-end/src/app/task.service.ts:
```diff
@@ -0,0 +1,26 @@
+    });
+  }
+
+  submitTask(id: number, title: string, comment: string) {
```
hosszú a paraméter lista, használj inkább 1 `Task` típusú paramétert

gy-szabo commented on front-end/src/app/app.component.ts:
```diff
@@ -1,35 +1,126 @@
+          ></app-task-card>
+        </div>
+        <div class="col-1"></div>
+        <div class="col-sm-5 done-container" *ngIf="!doneArrayEmpty">
```
itt van egy UX probléma: ha nagyon sok elem van a listában, akkor nagyon sokat kellene legörgetnem, hogy meg tudjam nyomni a `Clear` gombot. Több lehetőség is van ezt javítani:
1. a clear gombot a div tetjére teszed
2. beállítasz egy maximum magasságot, amit a div elfoglalhat, és ilyenkor görgethetővé teszed a listát

az utóbbi szofisztikáltabb, robosztusabb megoldás, mert pl ha lenne valamilyen content a listák alatt, akkor nem kéne az 1000 elemből álló done listát végig görgetnem

gy-szabo commented on front-end/src/app/task-card/task-card.component.html:
```diff
@@ -0,0 +1 @@
+<p>task-card works!</p>
```
a template-et inline kezeled, ezért erre a fájlra nincs szükség

gy-szabo commented on front-end/src/model/Task.ts:
```diff
@@ -1,9 +1,11 @@
 export class Task {
```
itt 2 probléma is van:
1. a deklarációknál csak a típus legyen megnevezve, értékadás ne töörténjen
1. az értékadás a konstuktorban kell megtörténjen

gy-szabo replied:
már majdnem jó, de default értékek se legyenek megadva, csak a típusokat add meg. Így gondolom:
```js
export class Task {
  id: number;
  // ...
```

bencefarkas200 committed:
"db.json" fájl törölve

bencefarkas200 committed:
Merge similar styles

In "app.component.css" put the common styles in one selector.

bencefarkas200 committed:
Delete unused imports

In "app.component.ts" unused imports were deleted.

bencefarkas200 committed:
Delete more unused imports

Deleted unused imports from "task.service.ts" and "task-card.component.ts".

bencefarkas200 committed:
Declare "taskService" in constructor

Instead of injecting "TaskService" at the begining, made an instance in the constructor in "app.component.ts".
(Also deleted some unused imports in "app.module.ts", but it won't matter once the component is standalone)

bencefarkas200 committed:
Use one "Task" type parameter

Made changes in "app.component.ts", "task.service.ts" and "Task.ts", so "task.service.ts" ' s submitTask function
uses one Task type parameter instead of three "simple".

bencefarkas200 committed:
Delete "task-card.component.html"

bencefarkas200 committed:
Put "Clear" button on top of done list

Also added scrolling to the lists, so they don't take up much space.

bencefarkas200 committed:
Make components standalone

bencefarkas200 committed:
Comment won't dissapear when the user tries to add new task without a title
Delete "javitando.txt"
(Also tried to implement snackbar, not working yet)

bencefarkas200 committed:
Change boolean variable names, so they describe better what they're for

bencefarkas200 committed:
Rearrange order of constructors and functions in components

bencefarkas200 committed:
Initialize "Task" class variables only in it's constructor

bencefarkas200 committed:
Snackbar now working

Also updated Angular version to 17.0.0.

bencefarkas200 committed:
Delete unused imports
